File: datamodel/src/main/java/datamodel/clinical/Medication.java
package datamodel.clinical;

import java.time.LocalDate;
import java.util.List;
import java.util.Set;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class Medication {

    @NotNull
    public abstract String name();

    @NotNull
    public abstract Set<String> categories();

    @Nullable
    public abstract MedicationStatus status();

    @Nullable
    public abstract String administrationRoute();

    @NotNull
    public abstract Dosage dosage();

    @Nullable
    public abstract LocalDate startDate();

    @Nullable
    public abstract LocalDate stopDate();

    @NotNull
    public abstract List<CypInteraction> cypInteractions();

    public abstract QTProlongatingRisk qtProlongatingRisk();

    @Nullable
    public abstract AtcClassification atc();
}

File: algo/src/main/kotlin/com/hartwig/actin/datamodel/clinical/Medication.kt
package com.hartwig.actin.datamodel.clinical

import com.hartwig.actin.datamodel.clinical.treatment.Drug
import java.time.LocalDate

data class Medication(
    val name: String,
    val status: MedicationStatus? = null,
    val administrationRoute: String? = null,
    val dosage: Dosage,
    val startDate: LocalDate? = null,
    val stopDate: LocalDate? = null,
    val drug: Drug? = null,
    val cypInteractions: List<CypInteraction> = emptyList(),
    val qtProlongatingRisk: QTProlongatingRisk,
    val atc: AtcClassification? = null,
    val isSelfCare: Boolean,
    val isTrialMedication: Boolean
) {
    
    fun allLevels(): Set<AtcLevel> {
        return atc?.allLevels() ?: emptySet()
    }
}
